File: tools/docs-generator/lib/level/engine/entity/entity_prefab_xml_reader.py
import xml.etree.ElementTree as EXml

from lib.level.engine.entity.component.component_prefab_xml_reader import ComponentPrefabXmlReader
from lib.level.engine.entity.entity_prefab import EntityPrefab
from lib.level.preload.preload_data import PreloadData
from lib.level.preload.preload_data_xml_reader import PreloadDataXmlReader


class EntityPrefabXmlReader(object):
    def __init__(self, preload_data_xml_reader, component_prefab_xml_reader):
        self.preload_data_xml_reader = preload_data_xml_reader
        self.component_prefab_xml_reader = component_prefab_xml_reader

    def read_prefab_from_file(self, filename):
        try:
            tree = EXml.parse(filename)
            element = tree.getroot()
            preload_data = PreloadData()

            if element.find(PreloadDataXmlReader.PRELOAD_DATA_ELEMENT) is not None:
                preload_data = self.preload_data_xml_reader.read_from_xml(
                    EXml.tostring(element.find(PreloadDataXmlReader.PRELOAD_DATA_ELEMENT))
                )

            component_prefabs = self.component_prefab_xml_reader.read_prefabs_from_string(
                EXml.tostring(element.find(ComponentPrefabXmlReader.COMPONENTS_ELEMENT))
            )

            return EntityPrefab(preload_data, component_prefabs)
        except Exception as e:
            raise EntityPrefabXmlReadException(filename, e)


class EntityPrefabXmlReadException(Exception):
    MESSAGE_TEMPLATE = "Cannot read entity prefab from file {}. Cause: {}."

    def __init__(self, filename, cause):
        super().__init__(self.MESSAGE_TEMPLATE.format(filename, cause))

File: core/src/remove/tanks/game/level/engine/entity/EntityPrefabXmlReader.java
package remove.tanks.game.level.engine.entity;

import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.utils.XmlReader;
import remove.tanks.game.level.engine.entity.component.ComponentPrefab;
import remove.tanks.game.level.engine.entity.component.ComponentPrefabXmlReader;
import remove.tanks.game.level.preload.PreloadData;
import remove.tanks.game.level.preload.PreloadDataXmlReader;

import java.util.ArrayList;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * @author Mateusz Długosz
 */
public final class EntityPrefabXmlReader {
    private final XmlReader xmlReader;
    private final PreloadDataXmlReader preloadDataXmlReader;
    private final ComponentPrefabXmlReader componentPrefabXmlReader;

    public EntityPrefabXmlReader(
            XmlReader xmlReader,
            PreloadDataXmlReader preloadDataXmlReader,
            ComponentPrefabXmlReader componentPrefabXmlReader
    ) {
        this.xmlReader = xmlReader;
        this.preloadDataXmlReader = preloadDataXmlReader;
        this.componentPrefabXmlReader = componentPrefabXmlReader;
    }

    public EntityPrefab readEntityPrefab(FileHandle fileHandle) {
        try {
            XmlReader.Element element = xmlReader.parse(fileHandle);
            return new EntityPrefab(
                    readPreloadData(element),
                    readComponents(element)
            );
        } catch (Exception e) {
            throw new EntityPrefabXmlReadException(fileHandle.path(), e);
        }
    }

    private PreloadData readPreloadData(XmlReader.Element element) {
        if (element.getChildByName(PreloadDataXmlReader.PRELOAD_DATA_ELEMENT) == null) {
            return new PreloadData(new ArrayList<>(), new ArrayList<>());
        }
        return preloadDataXmlReader.readPreloadData(element.getChildByName(PreloadDataXmlReader.PRELOAD_DATA_ELEMENT));
    }

    private Map<Class<? extends ComponentPrefab>, ComponentPrefab> readComponents(XmlReader.Element element) {
        return componentPrefabXmlReader.readComponentsPrefabs(
                element.getChildByName(ComponentPrefabXmlReader.COMPONENTS_ELEMENT)
        ).stream().collect(Collectors.toMap(ComponentPrefab::getClass, c -> c));
    }
}
